Update hello world to 0.12.0

jbesraa opened the pull request.

jbesraa committed:
update hello world tutorial

jbesraa committed:
update blog post

notmandatory commented on docs/_blog/hello-world.md:
```diff
@@ -89,12 +92,17 @@ since that's available out of the box in BDK and is pretty fast.
+    bitcoin::Network,
+    database::MemoryDatabase,
+    electrum_client::Client,
+    Wallet,
```
Need to include the `AddressIndex` import
```suggestion
   wallet::AddressIndex, 
    Wallet,
```

notmandatory commented on docs/_blog/hello-world.md:
```diff
@@ -131,21 +139,24 @@ Right now we will not get into details of all the available options in `TxBuilde
+    .fee_rate(FeeRate::from_sat_per_vb(5.0))
+    .do_not_spend_change()
+    .enable_rbf();
+let (mut psbt, tx_details) = tx_builder.finish()?
```
Missing semicolon.
```suggestion
let (mut psbt, tx_details) = tx_builder.finish()?;
```

notmandatory commented on docs/_blog/hello-world.md:
```diff
@@ -131,21 +139,24 @@ Right now we will not get into details of all the available options in `TxBuilde
+tx_builder
+    .add_recipient(faucet_address.script_pubkey(), 0_900)
+    .fee_rate(FeeRate::from_sat_per_vb(5.0))
+    .do_not_spend_change()
```
For this example to work you'll need to not disable spending change, and this is a less common thing to do so don't need to show it in this simple example.
```suggestion    
```

notmandatory commented on docs/_blog/hello-world.md:
```diff
@@ -131,21 +139,24 @@ Right now we will not get into details of all the available options in `TxBuilde
+
+let mut tx_builder = wallet.build_tx();
+tx_builder
+    .add_recipient(faucet_address.script_pubkey(), 0_900)
```
This should match the text of the post which says it's sending half the current balance.
```suggestion
    .add_recipient(faucet_address.script_pubkey(), balance / 2)
```

notmandatory commented on docs/_blog/hello-world.md:
```diff
@@ -131,21 +139,24 @@ Right now we will not get into details of all the available options in `TxBuilde
+let mut tx_builder = wallet.build_tx();
+tx_builder
+    .add_recipient(faucet_address.script_pubkey(), 0_900)
+    .fee_rate(FeeRate::from_sat_per_vb(5.0))
```
This line should also be removed since the text of the article assumes the default fee of 1 sat/vbyte is used.
```suggestion
```

notmandatory reviewed (changes requested):
I tested the code and found a few spots that need to be fixed to compile or to match the existing text. But once those are cleaned up this should be ready to go!

notmandatory commented on docs/tutorials/hello-world.md:
````diff
@@ -123,21 +131,24 @@ Right now we will not get into details of all the available options in `TxBuilde
 how to build transactions. We'll come back to this in a future article.
 
 ```rust
+use bdk::{bitcoin::Address, FeeRate};
````
Don't need FeeRate for this example.

```suggestion
use bdk::bitcoin::Address;
```

notmandatory commented on docs/tutorials/hello-world.md:
```diff
@@ -123,21 +131,24 @@ Right now we will not get into details of all the available options in `TxBuilde
+
+let mut tx_builder = wallet.build_tx();
+tx_builder
+    .add_recipient(faucet_address.script_pubkey(), 0_900)
```
To match text:
```suggestion
    .add_recipient(faucet_address.script_pubkey(), balance / 2)
```

notmandatory commented on docs/tutorials/hello-world.md:
```diff
@@ -123,21 +131,24 @@ Right now we will not get into details of all the available options in `TxBuilde
+tx_builder
+    .add_recipient(faucet_address.script_pubkey(), 0_900)
+    .fee_rate(FeeRate::from_sat_per_vb(5.0))
+    .do_not_spend_change()
```
Don't need this option and could make example fail if need to use change.
```suggestion
```

notmandatory commented on docs/tutorials/hello-world.md:
```diff
@@ -123,21 +131,24 @@ Right now we will not get into details of all the available options in `TxBuilde
+let mut tx_builder = wallet.build_tx();
+tx_builder
+    .add_recipient(faucet_address.script_pubkey(), 0_900)
+    .fee_rate(FeeRate::from_sat_per_vb(5.0))
```
To match text.
```suggestion
```

notmandatory commented on docs/tutorials/hello-world.md:
```diff
@@ -123,21 +131,24 @@ Right now we will not get into details of all the available options in `TxBuilde
+    .fee_rate(FeeRate::from_sat_per_vb(5.0))
+    .do_not_spend_change()
+    .enable_rbf();
+let (mut psbt, tx_details) = tx_builder.finish()?
```
Missing a semicolon.
```suggestion
let (mut psbt, tx_details) = tx_builder.finish()?;
```

notmandatory reviewed (changes requested):
The same changes from the blog need to also be made for the tutorial.

jbesraa committed:
add missing semicolon

jbesraa committed:
remove do not spend change

jbesraa committed:
change send amount to half balance

jbesraa committed:
remove fee rate

jbesraa committed:
add addressIndex

jbesraa committed:
fix tutrial to match the blog hello world

notmandatory committed:
Add link to latest docs and fix wallet.broadcast example

rajarshimaitra reviewed (commented):
tACK https://github.com/bitcoindevkit/bitcoindevkit.org/pull/63/commits/589e5bf577a7557cc42e2ce1f546aa81a5d83e48

Ran the code on bdk `v0.15.0` and all working. So I think the tutorial can even be  bumped to `v0.15.0`.

One surprising thing was the transaction actually broadcasted and there was testnet balance in the wallet. 
```
Wallet balance in SAT: 744846
Transaction Signed: true
Transaction details: TransactionDetails {
    transaction: None,
    txid: a5640812ba7091146d36427e61c17513354455a89e6af3984fc52dc40beea142,
    received: 297270,
    sent: 669837,
    fee: Some(
        144,
    ),
    confirmation_time: None,
    verified: true,
}
Transaction sent! TXID: a5640812ba7091146d36427e61c17513354455a89e6af3984fc52dc40beea142.
```

But now at each run the amount is draining to the hardcoded faucet address `mkHS9ne12qx9pS9VojpwU5xtRd4T7X7ZUt`. Maybe we can change the faucet address to an wallet address and keep rotating the fund?

notmandatory commented:
@rajarshimaitra good idea to rotate the testnet sats rather than send them back but I think that can be done later in a different PR.  The tutorial at least works now with the latest `bdk` so I'll merge it.

notmandatory committed:
Add note to hello-world blog to link to tutorial

Rather than keep updating this blog post now pointing readers to the tutorial page which we can keep updating with new bdk versions.
